File: leetcode/easy/valid_anagram.js
// https://leetcode.com/problems/valid-anagram/
// https://leetcode.com/submissions/detail/654303346/
// Time to write code: 1:53 min. (Original time to code: 15 min)
// O(n) time, O(n) space
// 3 companies asking this: Bloomberg, Microsoft, Facebook

let isAnagram = function(s, t) {
    if (s.length !== t.length) return false
    let freq = {}

    for (let i in s) {
        freq[s[i]] ? freq[s[i]]++ : freq[s[i]] = 1
        freq[t[i]] ? freq[t[i]]-- : freq[t[i]] = -1
    }

    for (let char of t) {
        if (freq[char] !== 0) return false
    }

    return true
}

File: leetcode/easy/valid_anagram.rb
# https://leetcode.com/problems/valid-anagram/
# Runtime: 72 ms, faster than 79.57% of Ruby online submissions for Valid Anagram.
# Memory Usage: 9.9 MB, less than 100.00% of Ruby online submissions for Valid Anagram.
def is_anagram(s, t)
    return false if s.size != t.size
    freq = Hash.new(0)
    s.size.times do |i|
        schar, tchar = s[i], t[i]
        freq[schar] += 1
        freq[tchar] -= 1
    end

    freq.each do |k, v|
        return false if v != 0
    end

    true
end
